### PR TITLE
[Experimental] Add mechanism to upgrade/downgrade between Add to Cart with Options block versions

### LINKED ISSUE #53208: [Enhancement]: Add mechanism to Upgrade/Downgrade the Blockified Add to Cart Experience
### Description
Add a way to allow users to upgrade or downgrade to the blockified `Add to Cart Form` experience. We can follow the same pattern we use to upgrade/downgrade from Products (Beta) to Product Collection block.

**Upgrade to**
Screenshot 2024-11-07 at 5 25 08 PM

**Revert to**
Screenshot 2024-11-07 at 5 27 30 PM

### Details

- **Functionality**
  - Provide a UI control for users to switch between the legacy and blockified Add to Cart Forms.
  - Handle any necessary cleanup or initialization when switching forms to prevent conflicts.

- **Implementation Considerations**
  - Ensure that the mechanism works across different product types and maintains compatibility with third-party extensions.

```[tasklist]
### Tasks
- [ ] Add the UI elements for upgrading/downgrading the Add to Cart Form.
- [ ] Test the mechanism thoroughly across various scenarios and product types.
```

### PR DESCRIPTION
**Pull request by Aljullu**
Closes #53208.

### How to test the changes in this Pull Request:

1. Install and activate the `WooCommerce Beta Tester` plugin
2. Go to `WooCommerce Admin Test Helper` -> `Features` and enable `blockified-add-to-cart` and `add-to-cart-with-options-stepper-layout`.
2. Create a post or page and add the Single Product block.
3. Select the existing _Add to Cart with Options_ block.
4. Verify in the sidebar there is a notice with a button to update. Click on it and verify it switches the block to the blockified version.
5. Do the same to switch back.
6. Try changing the display style to "Stepper" mode and verify it's preserved.

### CODE REVIEW

File: plugins/woocommerce-blocks/assets/js/blocks/product-elements/add-to-cart-form/components/upgrade-notice/index.tsx
Comment (kmanijak) [inline plugins/woocommerce-blocks/assets/js/blocks/product-elements/add-to-cart-form/components/upgrade-notice/index.tsx:101]: Can't help but notice the similarity between:
- plugins/woocommerce-blocks/assets/js/blocks/product-elements/add-to-cart-form/components/upgrade-notice/index.tsx
- plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/components/downgrade-notice/index.tsx
- https://github.com/woocommerce/woocommerce/blob/c7648abe5b5e1387c9e8ead887f1913b75e5443a/plugins/woocommerce-blocks/assets/js/blocks/product-query/inspector-controls/upgrade-notice.tsx#L9
- https://github.com/woocommerce/woocommerce/blob/c7648abe5b5e1387c9e8ead887f1913b75e5443a/plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/summary/upgrade.tsx#L38
- https://github.com/woocommerce/woocommerce/blob/c7648abe5b5e1387c9e8ead887f1913b75e5443a/plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/upgrade-notice.tsx#L33

For some of them I am of course guilty of repeating the same code 😅 Not the scope of this PR but we should consider extracting the notice as a separate block. Not expecting it in this PR but just mentioning for transparency.

File: plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/edit.tsx
Comment (kmanijak) [inline plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/edit.tsx:70]: In the current shape, if you add Add to Cart with Options (Experimental) block, you'd see the notice even though you didn't upgrade. In scope of Product Collection I used `convertedFromProducts` attribute to mark blocks upgraded from Products (Beta) and showed the notice only for them.

https://github.com/woocommerce/woocommerce/blob/c7648abe5b5e1387c9e8ead887f1913b75e5443a/plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/index.tsx#L253

add to cart

That's probably increases the scope of the whole feature but also we added couple of criteria to stop showing "Switch back" notice:

> // Upgrade Notice should be displayed only if:
> // - block is converted from Products
> // - user haven't acknowledged seeing the notice
> // - less than X hours since the notice was first displayed
> // - notice was displayed less than X times

https://github.com/woocommerce/woocommerce/blob/c7648abe5b5e1387c9e8ead887f1913b75e5443a/plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/index.tsx#L244C6-L262C3.

But considering the block is experimental I'll let you decide if that's blocking or not.

**Review by kmanijak — COMMENTED** [review]
PR is testing well! 🚀 

Based on experience in Product Collection block I shared a UX concern. But considering the block is experimental, if you don't plan to work on it at the moment, I'm happy to approve but would love to hear your opinion first!

File: plugins/woocommerce-blocks/assets/js/blocks/product-elements/add-to-cart-form/components/upgrade-notice/index.tsx
Comment (kmanijak) [inline plugins/woocommerce-blocks/assets/js/blocks/product-elements/add-to-cart-form/components/upgrade-notice/index.tsx]: After recent troubles we had with `createBlocksFromInnerBlocksTemplate` in various areas (https://github.com/woocommerce/woocommerce/issues/53765), I wanted to check how it behaves in Widgets and Patterns.

It seems the upgrade doesn't happen in widgets in classic theme and I think it's ok, because Single Product block doesn't work there anyway.

However, the event is recorded anyway. Considering the `upgradeToBlockifiedAddtoCartWithOptions` may not pass due to:

```
    if ( ! foundBlock ) {
		return;
	}
```

I think the `upgradeToBlockifiedAddtoCartWithOptions` should return boolean and the track should be recorded or not depends on the outcome:

```suggestion
	const handleClick = () => {
		const upgraded = upgradeToBlockifiedAddtoCartWithOptions( blockClientId );
		if ( upgraded ) {
			recordEvent( 'blocks_add_to_cart_with_options_migration', {
			    transform_to: 'blockified',
		    } );
		}
	};
```

It's minor and rather unlikely scenario but feels cleaner. The same would apply to downgrade.

**Review by octaedro — APPROVED** [review]
Good job @Aljullu! This is testing well on my end and the code LGTM 🚀 

For a follow-up PR (or a cooldown task), it might be a good idea to consolidate all the `upgrade` and `downgrade` notices into a single component for consistent use throughout the `WooCommerce-Blocks` project.

**UPDATE**
I just saw Karol's comments and reproduced what he mentioned. I think that we should address his comments before merging this PR.

File: plugins/woocommerce-blocks/assets/js/blocks/product-elements/add-to-cart-form/components/upgrade-notice/index.tsx
Comment (Aljullu) [reply plugins/woocommerce-blocks/assets/js/blocks/product-elements/add-to-cart-form/components/upgrade-notice/index.tsx:101]: Oh, good catch, I wasn't aware we had that many! Fully agree that we should unify them. :100:  I opened an issue here: https://github.com/woocommerce/woocommerce/issues/54076.

File: plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/edit.tsx
Comment (Aljullu) [reply plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/edit.tsx:70]: Good points! I'm a bit on the fence on this. Personally, I would prefer to leave it as "always visible" for now, at least until we have a release plan and we have investigated extension support.

If we release this block in an experimental state, I think it will be better to have the button always visible to switch to the stable block. What do you think? We currently do something similar with the "Revert to Classic Template" button in templates, it's always visible no matter if the user actively switched to the blockified template or not.

File: plugins/woocommerce-blocks/assets/js/blocks/product-elements/add-to-cart-form/components/upgrade-notice/index.tsx
Comment (Aljullu) [reply plugins/woocommerce-blocks/assets/js/blocks/product-elements/add-to-cart-form/components/upgrade-notice/index.tsx]: Good catch. We should probably also hide the block in the widgets screen. :grimacing: I will take a look at it.

For now, I implemented your suggestion in 50ce7c1c8834df7c32af375464da9f2812bfe5e2.

**Comment by Aljullu** [comment]
Thanks for the reviews, @kmanijak @octaedro. I fixed/answered the comments, would you mind taking another look? :bow:

File: plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/edit.tsx
Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/edit.tsx:70]: Yeah, I don't mind and it sounds ok! But wanted to mention this aspect anyway knowing we faced that earlier 🙌

File: plugins/woocommerce-blocks/assets/js/blocks/product-elements/add-to-cart-form/components/upgrade-notice/index.tsx
Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/blocks/product-elements/add-to-cart-form/components/upgrade-notice/index.tsx]: > Good catch. We should probably also hide the block in the widgets screen. 😬 I will take a look at it.

FWIW it can be hidden in here (wouldn't be surprised if you knew this already 😅 ):

https://github.com/woocommerce/woocommerce/blob/62b81a321a219f8e241e7f12ca9394fbd3aa9e97/plugins/woocommerce/src/Blocks/BlockTypesController.php#L521C6-L536C4

**Review by kmanijak — APPROVED** [review]
All changes look good to me! It tests well and as mentioned in comments, for now the notice will be displayed in every block and indefinitely no matter if it was transformed from previous block or not 👍 

Great job!

PS. Thanks for creating the issue to unify notices! 🚀

File: plugins/woocommerce-blocks/assets/js/blocks/product-elements/add-to-cart-form/components/upgrade-notice/index.tsx
Comment (Aljullu) [reply plugins/woocommerce-blocks/assets/js/blocks/product-elements/add-to-cart-form/components/upgrade-notice/index.tsx]: Thanks! I opened a PR here: https://github.com/woocommerce/woocommerce/pull/54087. Apparently switching from `registerBlockType()` to `registerProductBlockType()` was enough. Not sure if we should add the block to the PHP list as well, but considering we have https://github.com/woocommerce/woocommerce/issues/53691, I don't think it's needed. :thinking:

Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/blocks/product-elements/add-to-cart-form/components/upgrade-notice/index.tsx]: > considering we have https://github.com/woocommerce/woocommerce/issues/53691, I don't think it's needed. 🤔

Actually, I even picked it up today and PR is nearly ready (https://github.com/woocommerce/woocommerce/pull/54086) so definitely not worth the effort!

Comment (Aljullu) [reply plugins/woocommerce-blocks/assets/js/blocks/product-elements/add-to-cart-form/components/upgrade-notice/index.tsx]: Awesome! :heart_eyes: Feel free to ping me if you need a review!

**Pull request merged**
